### PR TITLE
Hook Up Monthly Contrib Checkout to Support Workers (Featuring Stripe)

### PR DESCRIPTION
**Pull request by Ap0c**
## Why are you doing this?

This wires the complete user journey for monthly contributions, for test users:

Monthly Contributions Checkout -> Client-Side Stripe -> Post to Support Workers -> Redirect to Thank You

**Note:** We don't have a thank you page to go with this yet, it's on the list.

[**Trello Card**](https://trello.com/c/lP5DwyH4/613-monthly-contributions-checkout-integrate-with-post-endpoint-currently-exists) and [**another**](https://trello.com/c/XxAOxZp3/633-integrate-test-users-with-the-monthly-contrib-checkout).

@davidfurey please go ahead and update the description with anything I've missed/correct anything I've explained poorly 🙂.

## Changes

- Changes the `/create` endpoint to only require a stripe token, not a user id.
- Temporarily made it so that only test users can access the checkout (this is still not ready for actual users to see, we don't want someone to come across it in production).
- Put user information into the guardian window object, so it can be accessed client-side (this only applies to authenticated routes).
- Added a fetch polyfill.
- Passed a callback to stripe, to be called when the token is ready.
- Refactored `TextInput` and `nameForm` components, to allow values to be read from, and send to, the redux store.
- Created client-side config for managing the user object, and sprinkled this information through the components and redux store as needed.
- Set up storage of currency in redux store.
- Stored checkout errors in the redux store.
- Created an `ajax` helper to post the checkout form (this might need a better name).
- Set the DEV `stage` to CODE to use the CODE support workers, and the DEV idapi endpoint to CODE (so we don't have to run either of these locally all the time).

## Screenshots

monthlycheckout

checkoutstripe

### CODE REVIEW

File: app/views/authenticatedFullUserReactTemplate.scala.html [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    <script type="text/javascript">
+        window.guardian = window.guardian || {};
+        window.guardian.user = {
+            email: "@user.primaryEmailAddress",
+            @user.publicFields.displayName.map { displayName =>
+            displayName: "@displayName",
+            }
+            @for(fields <- user.privateFields; firstName <- fields.firstName; lastName <- fields.secondName) {
+            firstName: "@firstName",
+            lastName: "@lastName",
```
Comment (svillafe) [inline app/views/authenticatedFullUserReactTemplate.scala.html]: Revisit in the future to set up a standard way of dealing with users inside the redux state.

File: app/views/authenticatedFullUserReactTemplate.scala.html [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            }
+        };
+    </script>
+    <script type="text/javascript" src="@assets(js)"></script>
+}
+
+@main(title = title, scripts = scripts) {
+    <main id="@id">
+        <noscript>
+            <div style="text-align: center;
```
Comment (svillafe) [inline app/views/authenticatedFullUserReactTemplate.scala.html]: If we are going to start to repeat the no js message in different pages, propably is better to put it inside a twirl fragment.

File: assets/components/textInput/textInput.jsx [hunk 3]
```diff
@@ -11,12 +11,14 @@ type PropTypes = {
...
   id?: string,
+  onChange?: (name: string) => void,
+  value?: string,
 };
 
 
 // ----- Functions ----- //
 
-function inputClass(hasLabel) {
+function inputClass(hasLabel: boolean): string {
```
Comment (svillafe) [inline assets/components/textInput/textInput.jsx]: Is it possible to add an Eslint rule to force that all the functions inside a file annotated with @flow must have type annotations?

File: app/views/authenticatedFullUserReactTemplate.scala.html (hunk 2 above)
Comment (Ap0c) [inline app/views/authenticatedFullUserReactTemplate.scala.html]: Agree, this is spread across three templates now.

On a related point, what is the difference between `authenticatedReact.scala.html` and `react.scala.html` @davidfurey?

File: app/lib/actions/ActionRefiners.scala [hunk 4]
```diff
@@ -37,7 +43,16 @@ class ActionRefiners(authenticatedIdUserProvider: Provider, idWebAppUrl: String,
     new AuthenticatedBuilder(authenticatedIdUserProvider, onUnauthenticated)
   }
 
+  private def authenticatedTestUser(onUnauthenticated: RequestHeader => Result = chooseRegister): ActionBuilder[AuthRequest] = {
+    new AuthenticatedBuilder(
+      authenticatedIdUserProvider.andThen(_.filter(user => testUsers.isTestUser(user.user.displayName))),
+      onUnauthenticated
+    )
+  }
```
Comment (rupertbates) [inline app/lib/actions/ActionRefiners.scala]: These curly braces are unneeded aren't they?

File: assets/pages/monthly-contributions/reducers/reducers.js [hunk 5]
```diff
@@ -8,16 +8,47 @@ import stripeCheckout from 'helpers/stripeCheckout/stripeCheckoutReducer';
...
+  action: Action): State {
 
   switch (action.type) {
 
     case 'SET_CONTRIB_VALUE':
-      return action.value;
+      return Object.assign({}, state, { amount: action.value });
+
+    case 'SET_FIRST_NAME':
+      return Object.assign({}, state, { firstName: action.name });
```
Comment (svillafe) [inline assets/pages/monthly-contributions/reducers/reducers.js]: Since we will have different forms that will accept first name and last name, I believe it would be better to extract the user logic from here and add it to a user reducer.

File: assets/helpers/config.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+// @flow
```
Comment (svillafe) [inline assets/helpers/config.js]: Why is this file called config? Is it not better to call it user since only cares about users data.

File: app/views/authenticatedFullUserReactTemplate.scala.html (hunk 2 above)
Comment (rupertbates) [inline app/views/authenticatedFullUserReactTemplate.scala.html]: A missing semi-colon from the non breaking space :-)

File: app/views/authenticatedFullUserReactTemplate.scala.html [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+                font-size: 30px;
+                padding-left: 20px;
+                padding-right: 20px;
+                background-color: #e9e939;">
+                Please enable JavaScript - we use it to provide the best experience for Guardian&nbspSupporters.<br/>
```
Comment (rupertbates) [inline app/views/authenticatedFullUserReactTemplate.scala.html]: As mentioned above there is a missing semi-colon in the nbsp in this sentence and the whole template looks like it could be dropped in favour of react.scala.html or vice versa.

File: assets/pages/monthly-contributions/reducers/reducers.js (hunk 5 above)
Comment (Ap0c) [inline assets/pages/monthly-contributions/reducers/reducers.js]: After discussion offline, we've decided 👍, we'll do this in a separate PR.

File: assets/helpers/config.js (hunk 6 above)
Comment (Ap0c) [inline assets/helpers/config.js]: I'd be happy with that, any preference @davidfurey?

File: assets/components/textInput/textInput.jsx [hunk 8]
```diff
@@ -26,18 +28,50 @@ function inputClass(hasLabel) {
...
+
+  const attrs: {
+    className: string,
+    id: ?string,
+    type: string,
+    placeholder: ?string,
+    value: string,
+    onChange?: Function,
+  } = {
+    className: inputClass(!!labelText),
```
Comment (rupertbates) [inline assets/components/textInput/textInput.jsx]: What does the !! do here?

File: assets/helpers/stripeCheckout/stripeCheckout.js [hunk 9]
```diff
@@ -26,15 +31,20 @@ const loadStripe = () => new Promise((resolve) => {
...
+  state: StripeState,
+  token: Function,
+  closed: Function,
+) => loadStripe().then(() => {
 
   stripeHandler = window.StripeCheckout.configure({
     name: 'Guardian',
     description: 'Please enter your card details.',
+    allowRememberMe: false,
     key: 'pk_test_Qm3CGRdrV4WfGYCpm0sftR0f',
```
Comment (rupertbates) [inline assets/helpers/stripeCheckout/stripeCheckout.js]: Should this key be here?

File: assets/components/textInput/textInput.jsx (hunk 8 above)
Comment (Ap0c) [inline assets/components/textInput/textInput.jsx]: Turns this into a boolean (JavaScript 🙃).

File: assets/helpers/stripeCheckout/stripeCheckout.js (hunk 9 above)
Comment (Ap0c) [inline assets/helpers/stripeCheckout/stripeCheckout.js]: This is the test public key ('publishable key') for the Stripe API, taken from [`membership-common`](https://github.com/guardian/membership-common/blob/master/conf/touchpoint.DEV.conf). In terms of privacy, it should be ok to version control it here, however it may be that there's a better place to keep it in the codebase in the interests of code cleanliness.

File: app/views/authenticatedFullUserReactTemplate.scala.html (hunk 2 above)
Comment (Ap0c) [inline app/views/authenticatedFullUserReactTemplate.scala.html]: 5fa212a

File: app/views/authenticatedFullUserReactTemplate.scala.html (hunk 7 above)
Comment (Ap0c) [inline app/views/authenticatedFullUserReactTemplate.scala.html]: Extra template dropped in 493e047. We're keeping this and the standard react one running in parallel because they do slightly different things. This is the non-cacheable, authenticated one that passes through user information to the client-side. The standard react one is cacheable and doesn't do anything with user info.

File: app/lib/actions/ActionRefiners.scala (hunk 4 above)
Comment (Ap0c) [inline app/lib/actions/ActionRefiners.scala]: f4ff04e

**Comment by Ap0c** [comment]
Ok, so we get a 500, an exception, on the `/create` POST.

**Comment by Ap0c** [comment]
The 500 was fixed in #89 (thanks @rupertbates), this looks good now.
